File: Dev/C++/Mascaret/src/VEHA/Kernel/Property.cpp
#include "VEHA/Kernel/Property.h"
#include "VEHA/Kernel/Class.h"
#include "VEHA/Kernel/ValueSpecification.h"
#include "VEHA/Kernel/DomainSpecification.h"
#include "VEHA/Kernel/Property.h"
#include "VEHA/Kernel/Classifier.h"

#include "VEHA/Kernel/LiteralReal.h"
#include "VEHA/Kernel/LiteralBoolean.h"
#include "VEHA/Kernel/LiteralString.h"
#include "VEHA/Kernel/LiteralInteger.h"
#include "VEHA/Kernel/InstanceValue.h"

namespace VEHA
{

Property::Property(const string& name,
		Class* cl,
		shared_ptr<Classifier> type,
		shared_ptr<Property> opposite,
		shared_ptr<ValueSpecification> def,
		shared_ptr<DomainSpecification> domain)
:Feature(name),
 _type(type), _classifier(cl), _opposite(opposite), _defaultValue(), _domain(domain),
 _isComposite(false), _isNavigable(false), _isReadOnly(false), _isDerived(false), _isUnique(false), _isOrdered(false),
 _minimumNumber(0), _maximumNumber(1)
{
	if (def) _defaultValue=def->clone();
	else
	{
		string typeName=type->getName();
		if (typeName=="real")
			_defaultValue=shared_ptr<ValueSpecification>(new LiteralReal());
		else if (typeName=="integer")
			_defaultValue=shared_ptr<ValueSpecification>(new LiteralInteger());
		else if (typeName=="string")
			_defaultValue=shared_ptr<ValueSpecification>(new LiteralString());
		else if (typeName=="bool")
			_defaultValue=shared_ptr<ValueSpecification>(new LiteralBoolean());
		else if (typeName=="boolean")
			_defaultValue=shared_ptr<ValueSpecification>(new LiteralBoolean());
		//else
		//	_defaultValue=shared_ptr<ValueSpecification>(new InstanceValue(_type));
	}

}

Property::~Property()
{
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Kernel/Property.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace Mascaret
{
    public class Property : Feature
    {

        private Classifier type = null;
        public Classifier Type
        {
            get { return type; }
            set { type = value; }
        }


        private Classifier classifier = null;
        public Classifier Classifier
        {
            get { return classifier; }
            set { classifier = value; }
        }


        private bool isComposite = false;
        public bool IsComposite
        {
            get { return isComposite; }
            set { isComposite = value; }
        }


        private bool isNavigable = false;
        public bool IsNavigable
        {
            get { return isNavigable; }
            set { isNavigable = value; }
        }


        private bool isReadOnly = false;
        public bool IsReadOnly
        {
            get { return isReadOnly; }
            set { isReadOnly = value; }
        }


        private bool isDerived = false;
        public bool IsDerived
        {
            get { return isDerived; }
            set { isDerived = value; }
        }


        private bool isOrdered = false;
        public bool IsOrdered
        {
            get { return isOrdered; }
            set { isOrdered = value; }
        }


        private bool isUnique = false;
        public bool IsUnique
        {
            get { return isUnique; }
            set { isUnique = value; }
        }


        private int minimumNumber = 0;
        public int MinimumNumber
        {
            get { return minimumNumber; }
            set { minimumNumber = value; }
        }


        private int maximumNumber = 1;
        public int MaximumNumber
        {
            get { return maximumNumber; }
            set { maximumNumber = value; }
        }


        private Property opposite = null;
        public Property Opposite
        {
            get { return opposite; }
            set { opposite = value; }
        }


        private ValueSpecification defaultValue = null;
        public ValueSpecification DefaultValue
        {
            get { return defaultValue; }
            set { defaultValue = value; }
        }


        private DomainSpecification domain = null;
        public DomainSpecification Domain
        {
            get { return domain; }
            set { domain = value; }
        }

        public override string getFullName()
        {
            return type.getFullName() + "." + name;
        }

        //Default Parameters opposite = null ,ValueSpecification def= ,null,DomainSpecification domain=null
        public Property(string name, Class cl, Classifier type, Property opposite, ValueSpecification def, DomainSpecification domain)
            : base(name)
        {
            this.type = type;
            this.classifier = cl;
            this.opposite = opposite;
            this.defaultValue = null;
            this.domain = domain;
            this.isComposite = false;
            this.isNavigable = false;
            this.isReadOnly = false;
            this.isDerived = false;
            this.isUnique = false;
            this.isOrdered = false;
            this.minimumNumber = 0;
            this.maximumNumber = 1;

            if (def != null)
                this.defaultValue = def;
            else
            {
                string typeName = type.name;
                if (typeName == "real")
                {
                    this.defaultValue = new LiteralReal();
                }
                else if (typeName == "integer")
                {
                    this.defaultValue = new LiteralInteger();
                }
                else if (typeName == "string")
                {
                    this.defaultValue = new LiteralString();
                }
                else if (typeName == "boolean")
                {
                    this.defaultValue = new LiteralBoolean();
                }
                else if (typeName == "bool")
                {
                    this.defaultValue = new LiteralBoolean();
                }
            }

        }

        public ValueSpecification createValueFromString(string str)
        {
           // MascaretApplication.Instance.logfile.WriteLine("Property::createValue : " + str); MascaretApplication.Instance.logfile.Flush();
            return type.createValueFromString(str);
        }

    }
}